### ISSUE TITLE
CosMx input file format requirements

### ISSUE THREAD

**Issue by Jwong684** [gustaveroussy/sopa#63]
Hi Sopa development team,

Sopa looks like a perfect fit for what I've been trying to do with integrating IF stains with Baysor re-segmentation. However, I've been having trouble figuring out how to run the data properly since the newest CosMx output data (and all the column names) are quite different from the example ones on their website. I'm not exactly sure what columns/root file names sopa is looking for. Do you have any documentation there?

I am currently trying to run your Snakefile. From your FAQ, I see that you need these three components:
data_path is the directory containing (i) the transcript file (ending with _tx_file.csv or _tx_file.csv.gz), (ii) the FOV locations file, and (iii) a Morphology2D directory containing the images.

I presumed that these files were based on this CosMx README: 
https://nanostring-public-share.s3.us-west-2.amazonaws.com/SMI-Compressed/SMI-ReadMe.html

The Morphology2D directory has files in this format:
```
20240215_023634_S3_C902_P99_N99_F001.TIF
20240215_023634_S3_C902_P99_N99_F002.TIF
```
for FOV001 and FOV002

The transcripts file looks like this:
```
,CellComp,CellId,Spot1_count,Spot2_count,Spot3_count,Spot4_count,codeclass,fov,multicolor_spots_per_feature,possible_BC_count,random_call_probability,seed_x,seed_y,spots_per_feature,target,target_call_observations,target_count_per_feature,target_idx,x,y,z
0,Cytoplasm,1,1,1,1,1,Endogenous,1,0,1,0.0025906,40612,14474,4,CD86,4,1,935,4061.07,1447.53,0
1,Nuclear,3,1,1,1,1,SystemControl,1,1,2,0.00517446,41620,14490,4,SystemControl100,4,1,537,4162.15,1449.18,0
2,Nuclear,2,1,2,1,2,Endogenous,1,0,1,0.0025906,42113,14517,4,RPL37,6,1,937,4211.38,1451.83,0
```
I'm not entirely sure what the FOV locations file is supposed to look like - there are x and y coordinates for each transcript and for each cell, though it seems like sopa is looking for one X and Y coordinate for each FOV if I'm understanding it right. 

I get errors like this:

```
rule to_spatialdata:
    input: sopa/data/fov001
    output: sopa/data/fov001.zarr/.zgroup
    jobid: 4
    reason: Missing output files: sopa/data/fov001.zarr/.zgroup
    resources: tmpdir=/tmp, mem_mb=128000, mem_mib=122071

Activating conda environment: ../../../envs/env_sopa
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│conda_environments/radian/lib/python3.10/site-packages/pandas/core/inde │
│ xes/base.py:3805 in get_loc                                                                      │
│                                                                                                  │
│   3802 │   │   """                                                                               │
│   3803 │   │   casted_key = self._maybe_cast_indexer(key)                                        │
│   3804 │   │   try:                                                                              │
│ ❱ 3805 │   │   │   return self._engine.get_loc(casted_key)                                       │
│   3806 │   │   except KeyError as err:                                                           │
│   3807 │   │   │   if isinstance(casted_key, slice) or (                                         │
│   3808 │   │   │   │   isinstance(casted_key, abc.Iterable)                                      │
│                                                                                                  │
│ ╭───────────────────────────────────────── locals ─────────────────────────────────────────╮     │
│ │ casted_key = 'X_mm'                                                                      │     │
│ │        key = 'X_mm'                                                                      │     │
│ │       self = Index(['Unnamed: 0', 'CellId', 'Spot1_count', 'Spot2_count', 'Spot3_count', │     │
│ │              │      'Spot4_count', 'codeclass', 'fov', 'multicolor_spots_per_feature',   │     │
│ │              │      'possible_BC_count', 'random_call_probability', 'seed_x', 'seed_y',  │     │
│ │              │      'spots_per_feature', 'target', 'target_call_observations',           │     │
│ │              │      'target_count_per_feature', 'target_idx', 'x', 'y', 'z'],            │     │
│ │              │     dtype='object')                                                       │     │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────╯     │
│                                                                                                  │
│ in pandas._libs.index.IndexEngine.get_loc:167                                                    │
│                                                                                                  │
│ in pandas._libs.index.IndexEngine.get_loc:196                                                    │
│                                                                                                  │
│ in pandas._libs.hashtable.PyObjectHashTable.get_item:7081                                        │
│                                                                                                  │
│ in pandas._libs.hashtable.PyObjectHashTable.get_item:7089                                        │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
KeyError: 'X_mm'
```
In this case, X_mm isn't a column header in the sample data or in the newest CosMx output.

Thank you for your help!

**Comment by quentinblampey** [comment]
Hi @Jwong684,

Indeed, it seems that the file format has changed. Currently, I'm using:
- The columns `x_global_px`, `y_global_px`, and `target` for the transcript file
- All the images inside the `Morphology2D` directory
- The columns `X_mm`, `Y_mm` and `FOV` for the FOV location file (see the example below)

This is an example of a FOV position file (i.e. a file with suffix `_fov_positions_file.csv`):
```csv
Slide,FOV,X_mm,Y_mm
3,1,0.722208817050781,13.7198104736064
3,2,1.23412451267148,13.7198104736064
3,3,1.74604020829218,13.7198104736064
...
```

I'm trying to find some public data with the new format (or at least their official file description) and I'll work on an update of the CosMX reader. I'll let you know!

**Comment by Jwong684** [comment]
Thank you Quentin, that was helpful. I tried to reformat and rename some of my columns, but I'm getting this bizarre KeyError.

```
rule to_spatialdata:
    input: /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001
    output: /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001.zarr/.zgroup
    jobid: 4
    reason: Missing output files: /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001.zarr/.zgroup
    resources: tmpdir=/tmp, mem_mb=128000, mem_mib=122071

Activating conda environment: ../../../envs/env_sopa
[WARNING] (sopa.io.reader.cosmx) Image stitching is currently experimental
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /projects/rmorin_scratch/conda_environments/radian/lib/python3.10/site-packages/sopa/cli/app.py: │
│ 98 in read                                                                                       │
│                                                                                                  │
│    95 │   │   io, technology                                                                     │
│    96 │   ), f"Technology {technology} unknown. Currently available: xenium, merscope, cosmx,    │
│    97 │                                                                                          │
│ ❱  98 │   sdata = getattr(io, technology)(data_path, **kwargs)                                   │
│    99 │   io.write_standardized(sdata, sdata_path, delete_table=True)                            │
│   100                                                                                            │
│   101                                                                                            │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │ config_path = None                                                                           │ │
│ │   data_path = '/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001/'              │ │
│ │          io = <module 'sopa.io' from                                                         │ │
│ │               '/projects/rmorin_scratch/conda_environments/radian/lib/python3.10/site-packa… │ │
│ │      kwargs = {}                                                                             │ │
│ │        Path = <class 'pathlib.Path'>                                                         │ │
│ │  sdata_path = PosixPath('/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001.zar… │ │
│ │  technology = 'cosmx'                                                                        │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /projects/rmorin_scratch/conda_environments/radian/lib/python3.10/site-packages/sopa/io/reader/c │
│ osmx.py:68 in cosmx                                                                              │
│    65 │   ### Read image(s)                                                                      │
│    66 │   images_dir = _find_dir(path, "Morphology2D")                                           │
│    67 │   if fov is None:                                                                        │
│ ❱  68 │   │   image, protein_names = _read_stitched_image(                                       │
│    69 │   │   │   images_dir,                                                                    │
│    70 │   │   │   fov_locs,                                                                      │
│    71 │   │   │   protein_dir_dict,                                                              │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │          dataset_id = '/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001/fov00… │ │
│ │                 fov = None                                                                   │ │
│ │              fov_id = None                                                                   │ │
│ │            fov_locs = │    Slide  x_global_px  y_global_px   X_mm   Y_mm          xmin  xmax │ │
│ │                       ymin         ymax                                                      │ │
│ │                       FOV                                                                    │ │
│ │                       1        3      15.6454     4.296388  2.816  0.773  23411.854637     0 │ │
│ │                       0  6426.620609                                                         │ │
│ │ image_models_kwargs = {'chunks': (1, 1024, 1024), 'scale_factors': [2, 2, 2, 2]}             │ │
│ │          images_dir = PosixPath('/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fo… │ │
│ │       imread_kwargs = {}                                                                     │ │
│ │                path = PosixPath('/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fo… │ │
│ │    protein_dir_dict = {}                                                                     │ │
│ │       read_proteins = False                                                                  │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /projects/rmorin_scratch/conda_environments/radian/lib/python3.10/site-packages/sopa/io/reader/c │
│ osmx.py:182 in _read_stitched_image                                                              │
│                                                                                                  │
│   179 │   pattern = re.compile(r".*_F(\d+)")                                                     │
│   180 │   for image_path in images_dir.iterdir():                                                │
│   181 │   │   if image_path.suffix == ".TIF":                                                    │
│ ❱ 182 │   │   │   fov = int(pattern.findall(image_path.name)[0])                                 │
│   183 │   │   │                                                                                  │
│   184 │   │   │   image, protein_names = _read_fov_image(                                        │
│   185 │   │   │   │   image_path, protein_dir_dict.get(fov), **imread_kwargs                     │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │       fov_images = {}                                                                        │ │
│ │         fov_locs = │    Slide  x_global_px  y_global_px   X_mm   Y_mm          xmin  xmax    │ │
│ │                    ymin         ymax                                                         │ │
│ │                    FOV                                                                       │ │
│ │                    1        3      15.6454     4.296388  2.816  0.773  23411.854637     0    │ │
│ │                    0  6426.620609                                                            │ │
│ │       image_path = PosixPath('/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov00… │ │
│ │       images_dir = PosixPath('/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov00… │ │
│ │    imread_kwargs = {}                                                                        │ │
│ │          pattern = re.compile('.*_F(\\d+)')                                                  │ │
│ │ protein_dir_dict = {}                                                                        │ │
│ │    protein_names = None                                                                      │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
IndexError: list index out of range
```

My files look like this:
```
==> fov001_fov_positions_file.csv <==
Slide,FOV,x_global_px,y_global_px,X_mm,Y_mm
3,1,15.645400102017447,4.296387956244871,2.816,0.773

==> fov001_tx_file.csv <==
fov,CellId,x_global_px,y_global_px,target
1,1,4061.07,1447.53,CD86
1,3,4162.15,1449.18,SystemControl100
1,2,4211.38,1451.83,RPL37
1,2,4231.33,1453.2,HDAC1
1,0,3940.3,1453.38,IGHG2
1,2,4228.43,1454.57,RPL37
1,2,4213.23,1455.55,GPX1
1,1,4068.62,1455.55,MYL12A
1,2,4229.48,1456.5,IGHG1
```
Notably, I don't think there is a x_global_px and y_global_px metric in the output of CosMx transcript data. They only have x and y coordinates for a given FOV. Otherwise, they have a x_global_px and y_global_px for a cell that was determined by cellpose.

**Comment by cstrlln** [comment]
Wanted to add that for getting the data in similar format to previous datasets, it has to be exported as a flatfile in AtoMX.

**Comment by quentinblampey** [comment]
Nice @cstrlln! So the format didn't change? Is it just a matter of using one specific export module?

**Comment by cstrlln** [comment]
I believe it is the same format. The problem with CosMX data output is that is very poorly documented, and there might be some little changes here and there, as discussed here also in proseg: https://github.com/dcjones/proseg/issues/3

**Comment by Jwong684** [comment]
Thanks @cstrlln! I agree that the Cosmx data output is quite poorly documented.

If I just export the flatfiles from their AtoMx server, I can get output that resembles the example data.

tx_file.csv.gz 
```
fov,cell_ID,cell,x_local_px,y_local_px,x_global_px,y_global_px,z,target,CellComp
1,0,c_1_1_0,44,25,2155.99997401111,97926.9987946669,7,APOA1,None
1,0,c_1_1_0,28,51,2139.99997401111,97900.9987946669,7,MMP14,None
1,0,c_1_1_0,53,63,2164.99997401111,97888.9987946669,7,IGHG2,None
1,0,c_1_1_0,97,84,2208.99997401111,97867.9987946669,7,STMN1,None
```
fov_positions_file.csv.gz
```
Slide,X_mm,Y_mm,Z_mm,ZOffset_mm,ROI,FOV,Order,Run_Tissue_name
1,0.254033352714032,11.7817589796614,0.022991,-2,0,1,0,slide1
1,0.767873543431051,11.7836834747577,0.022962,-2,0,2,7,slide1
1,1.43182435166091,11.8568142884178,0.022917,-2,0,3,8,slide1
```

> Also, according to the error, it seems that one image doesn't end with _F<N>.TIF, where <N> is the FOV number. Can you check that all TIF images have the right suffix?
I was modifying the file name to see what sopa was looking for. It does originally end with a _F001.TIF for fov001.

Having done that, I still get an key error about X_mm:
```
Activating conda environment: .snakemake/conda/7fa3da0bba643754f940be8428f9d977_
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/.snakemake/conda/7fa3da0bba643754f940be8428f9 │
│ d977_/lib/python3.10/site-packages/pandas/core/indexes/base.py:3805 in get_loc                   │
│                                                                                                  │
│   3802 │   │   """                                                                               │
│   3803 │   │   casted_key = self._maybe_cast_indexer(key)                                        │
│   3804 │   │   try:                                                                              │
│ ❱ 3805 │   │   │   return self._engine.get_loc(casted_key)                                       │
│   3806 │   │   except KeyError as err:                                                           │
│   3807 │   │   │   if isinstance(casted_key, slice) or (                                         │
│   3808 │   │   │   │   isinstance(casted_key, abc.Iterable)                                      │
│                                                                                                  │
│ ╭───────────────────────────────────── locals ──────────────────────────────────────╮            │
│ │ casted_key = 'X_mm'                                                               │            │
│ │        key = 'X_mm'                                                               │            │
│ │       self = Index(['Slide', 'Y_mm', 'Z_mm', 'ZOffset_mm', 'ROI', 'FOV', 'Order', │            │
│ │              │      'Run_Tissue_name'],                                           │            │
│ │              │     dtype='object')                                                │            │
│ ╰───────────────────────────────────────────────────────────────────────────────────╯            │
│                                                                                                  │
│ in pandas._libs.index.IndexEngine.get_loc:167                                                    │
│                                                                                                  │
│ in pandas._libs.index.IndexEngine.get_loc:196                                                    │
│                                                                                                  │
│ in pandas._libs.hashtable.PyObjectHashTable.get_item:7081                                        │
│                                                                                                  │
│ in pandas._libs.hashtable.PyObjectHashTable.get_item:7089                                        │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
KeyError: 'X_mm'
The above exception was the direct cause of the following exception:

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/.snakemake/conda/7fa3da0bba643754f940be8428f9 │
│ d977_/lib/python3.10/site-packages/sopa/cli/app.py:98 in read                                    │
│                                                                                                  │
│    95 │   │   io, technology                                                                     │
│    96 │   ), f"Technology {technology} unknown. Currently available: xenium, merscope, cosmx,    │
│    97 │                                                                                          │
│ ❱  98 │   sdata = getattr(io, technology)(data_path, **kwargs)                                   │
│    99 │   io.write_standardized(sdata, sdata_path, delete_table=True)                            │
│   100                                                                                            │
│   101                                                                                            │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │ config_path = None                                                                           │ │
│ │   data_path = '/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001/'              │ │
│ │          io = <module 'sopa.io' from                                                         │ │
│ │               '/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/.snakemake/conda/7fa3da0b… │ │
│ │      kwargs = {}                                                                             │ │
│ │        Path = <class 'pathlib.Path'>                                                         │ │
│ │  sdata_path = PosixPath('/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001.zar… │ │
│ │  technology = 'cosmx'                                                                        │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/.snakemake/conda/7fa3da0bba643754f940be8428f9 │
│ d977_/lib/python3.10/site-packages/sopa/io/reader/cosmx.py:54 in cosmx                           │
│                                                                                                  │
│    51 │   image_models_kwargs, imread_kwargs = _default_image_kwargs(image_models_kwargs, imre   │
│    52 │                                                                                          │
│    53 │   dataset_id = _infer_dataset_id(path, dataset_id)                                       │
│ ❱  54 │   fov_locs = _read_cosmx_fov_locs(path, dataset_id)                                      │
│    55 │   fov_id, fov = _check_fov_id(fov)                                                       │
│    56 │                                                                                          │
│    57 │   protein_dir_dict = {}                                                                  │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │          dataset_id = '/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001/fov00… │ │
│ │                 fov = None                                                                   │ │
│ │ image_models_kwargs = {'chunks': (1, 1024, 1024), 'scale_factors': [2, 2, 2, 2]}             │ │
│ │       imread_kwargs = {}                                                                     │ │
│ │                path = PosixPath('/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fo… │ │
│ │       read_proteins = False                                                                  │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/.snakemake/conda/7fa3da0bba643754f940be8428f9 │
│ d977_/lib/python3.10/site-packages/sopa/io/reader/cosmx.py:163 in _read_cosmx_fov_locs           │
│                                                                                                  │
│   160 │                                                                                          │
│   161 │   pixel_size = 0.120280945  # size of a pixel in microns                                 │
│   162 │                                                                                          │
│ ❱ 163 │   fov_locs["xmin"] = fov_locs["X_mm"] * 1e3 / pixel_size                                 │
│   164 │   fov_locs["xmax"] = 0  # will be filled when reading the images                         │
│   165 │                                                                                          │
│   166 │   fov_locs["ymin"] = 0  # will be filled when reading the images                         │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │ dataset_id = '/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001/fov001'         │ │
│ │   fov_file = PosixPath('/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001/fov0… │ │
│ │   fov_locs = │   │     Slide       Y_mm      Z_mm  ZOffset_mm  ROI  FOV  Order               │ │
│ │              Run_Tissue_name                                                                 │ │
│ │              X_mm                                                                            │ │
│ │              0.254033      1  11.781759  0.022991          -2    0    1      0  CT20-10 A1   │ │
│ │              LLMPP P01 PBL TMA-1                                                             │ │
│ │       path = PosixPath('/projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001')     │ │
│ │ pixel_size = 0.120280945                                                                     │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/.snakemake/conda/7fa3da0bba643754f940be8428f9 │
│ d977_/lib/python3.10/site-packages/pandas/core/frame.py:4102 in __getitem__                      │
│                                                                                                  │
│    4099 │   │   if is_single_key:                                                                │
│    4100 │   │   │   if self.columns.nlevels > 1:                                                 │
│    4101 │   │   │   │   return self._getitem_multilevel(key)                                     │
│ ❱  4102 │   │   │   indexer = self.columns.get_loc(key)                                          │
│    4103 │   │   │   if is_integer(indexer):                                                      │
│    4104 │   │   │   │   indexer = [indexer]                                                      │
│    4105 │   │   else:                                                                            │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │         is_mi = False                                                                        │ │
│ │ is_single_key = True                                                                         │ │
│ │           key = 'X_mm'                                                                       │ │
│ │          self = │   │     Slide       Y_mm      Z_mm  ZOffset_mm  ROI  FOV  Order            │ │
│ │                 Run_Tissue_name                                                              │ │
│ │                 X_mm                                                                         │ │
│ │                 0.254033      1  11.781759  0.022991          -2    0    1      0  CT20-10   │ │
│ │                 A1 LLMPP P01 PBL TMA-1                                                       │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/.snakemake/conda/7fa3da0bba643754f940be8428f9 │
│ d977_/lib/python3.10/site-packages/pandas/core/indexes/base.py:3812 in get_loc                   │
│                                                                                                  │
│   3809 │   │   │   │   and any(isinstance(x, slice) for x in casted_key)                         │
│   3810 │   │   │   ):                                                                            │
│   3811 │   │   │   │   raise InvalidIndexError(key)                                              │
│ ❱ 3812 │   │   │   raise KeyError(key) from err                                                  │
│   3813 │   │   except TypeError:                                                                 │
│   3814 │   │   │   # If we have a listlike key, _check_indexing_error will raise                 │
│   3815 │   │   │   #  InvalidIndexError. Otherwise we fall through and re-raise                  │
│                                                                                                  │
│ ╭───────────────────────────────────── locals ──────────────────────────────────────╮            │
│ │ casted_key = 'X_mm'                                                               │            │
│ │        key = 'X_mm'                                                               │            │
│ │       self = Index(['Slide', 'Y_mm', 'Z_mm', 'ZOffset_mm', 'ROI', 'FOV', 'Order', │            │
│ │              │      'Run_Tissue_name'],                                           │            │
│ │              │     dtype='object')                                                │            │
│ ╰───────────────────────────────────────────────────────────────────────────────────╯            │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
KeyError: 'X_mm'

Error in rule to_spatialdata:
    jobid: 4
    input: /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001
    output: /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001.zarr/.zgroup
    conda-env: /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/.snakemake/conda/7fa3da0bba643754f940be8428f9d977_
    shell:

        sopa read /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001/ --sdata-path /projects/rmorin_scratch/jawong_tmp/cosmx/pbl/sopa/data/fov001.zarr --technology "cosmx"

```

**Comment by Jwong684** [comment]
Ah I just saw your other thread about this: https://github.com/gustaveroussy/sopa/issues/65

It might be good to add a parameter to set an ID name for the data. I guess at the moment it just matches anything preceding *_tx_file.csv and *_fov_positions_file.csv as the dataset ID

**Comment by cstrlln** [comment]
> Ah I just saw your other thread about this: #65
> 
> It might be good to add a parameter to set an ID name for the data. I guess at the moment it just matches anything preceding *_tx_file.csv and *_fov_positions_file.csv as the dataset ID

@Jwong684 The issue right now is that the columns are read in a hardcoded order based on previous versions of CosMX output. The flatfiles should work otherwise.

**Comment by quentinblampey** [comment]
Indeed @Jwong684, as mentioned in this issue, I was using the second CSV column as the pandas index. But it seems that the FOV column is not always the second column. I have a fix on the `dev` branch if you want to try it out, and I will soon release a new version of sopa.

> It might be good to add a parameter to set an ID name for the data. I guess at the moment it just matches anything preceding *_tx_file.csv and *_fov_positions_file.csv as the dataset ID

Actually we first infer the `dataset_id` using the fov_positions_file.csv, and then we keep this ID to detect the other files (instead of simply looking for all files that match this pattern). You can also provide `dataset_id` as an argument to the reader if you don't want it to be inferred.

**Comment by quentinblampey** [comment]
This should be fixed in `sopa==1.0.12`, can you try?

**Comment by quentinblampey** [comment]
This is fixed, and the docs are updated. Closing the issue :)
